Add planner for dynamic pricing

schenlap opened the pull request:
- [x] move `cheap` rate to site
- [x] remove tariff `IsCheap` api
- [x] add tests for planner
- [ ] integrate with existing planner

schenlap committed:
swap paramter of AssumedChargeDuration

schenlap committed:
add dynamic pricing for awattar

schenlap committed:
add dynamic pricing for awattar

schenlap committed:
rebase to master

schenlap committed:
further improvments

andig commented:
Ich glaub ich habs grob verstanden. Eine Pricetable generiert einen Plan über n nicht-kontinuierliche Slots und startet/stoppt bei jedem Planslot (außer dem letzten wo nicht mehr gestoppt wird). 
Der letzte müsste sich an der Zielzeit bemessen, nicht daran ob er der letzte des Plans ist?

Wenn das so grob stimmt dann könnten wir mal versuchen das zu zerlegen (und ich helfe gerne):
- Tarife generieren `PriceTables`
- PriceTable ist sortierbar ;)
- Site verwurstet PriceTable mit einem `Planner`
- Planner wird anstatt `isCheap` and Loadpoint übergeben und signalisiert oder zu laden ist (`Active()`?)

andig commented:
Übel wirds wenn Lastmanagement kommt, aber das ist eine ganz andere Geschichte…

andig committed:
Remove comment

andig committed:
test commit

schenlap commented:
Ja, das hast du richtig verstanden.

> Der letzte müsste sich an der Zielzeit bemessen, nicht daran ob er der letzte des Plans ist?

beides ist möglich. Es geht aber nicht um Zielzeit sondern um erwartet ladezeit. Der letzte Slot kann auch Stunden vor der Zielzeit liegen. Beim Testen muss man aufpassen, Awattar sperrt nach zu vielen Versuchen die API (vermutlich für die IP Adresse), daher hatte ich das fake_awattar drinnen.

andig committed:
Format

andig committed:
Move logic to planner component

andig committed:
Move sorting to planner

andig committed:
wip

andig committed:
wip

andig commented:
@schenlap ich habe mal angefangen aufzuräumen. Die Logik gibts jetzt nur noch 1x in einer neuen Planner Komponente. Ein paar Todos hab ich oben ergänzt.

andig committed:
wip

andig commented on core/loadpoint_api.go:
```diff
@@ -245,3 +244,13 @@ func (lp *LoadPoint) GetRemainingEnergy() float64 {
 	return lp.chargeRemainingEnergy
 }
+
+// GetFinishAt returns the charge finish time
```
warum brauchst du beide Funktionen? Sind die nicht redundant?

schenlap replied:
nö. Das eine sagt wenn das Auto geladen haben soll (Zielzeit), getAssumedDuration, sagt wie lange das Laden dauert.

andig replied:
Ok, dann sollte das analog setter

    GetTargetTime

Heissen

schenlap replied:
erledigt

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+func NewPlanner(log *util.Logger, tariff api.Tariff) *Planner {
+	return &Planner{
+		log:    log,
+		cheap:  0.2, // TODO
```
muss aus der site kommen

schenlap replied:
ich würde das cheap rausschmeißen. hat keinen Sinn mehr. Wir laden wenn es am günstigsten ist und aus.

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+	t.mux.Lock()
+	defer t.mux.Unlock()
+
+	if end.Before(time.Now()) {
```
sollte die Abfrage nicht außerhalb passieren?

schenlap replied:
hier wird abgefragt ob wir überhaupt eine Preisinfo für die Zukunft haben? Denke das passt hier.

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+		return false, nil
+	}
+
+	duration = time.Duration(float64(duration) * 1.05) // increase by 5%
```
warum?

schenlap replied:
Anpassung der geschätzen Zeit an die tatsächliche Zeit. Beim Ziellladen haben wir hier 20%, sollten wir dann vereinheitlichen

andig replied:
Denke auch. Das ist eine globale Annahme die const sein sollte

andig replied:
Const chargeEfficiency = 0.95

Haben wir ähnlich auch im ChargeTimer- einmal sollte reichen und beide die gleiche Konstante verwenden.

schenlap replied:
ja, sehe ich auch so. Ich kann darauf aber nicht zugreifen, ist in einem anderen package.

andig replied:
Einfach „public“ machen- großer Anfangsbuchstabe. Dann kannst Du die auch nutzen.

schenlap replied:
nö, so einfach ist es nicht. Ich müsste vermutlich auch das soc package importieren. Ist auch nicht schön hier eine weitere Abhängigkeit dazu zu bekommen. Ev. sollte es dann wo anders definiert werden?

andig replied:
Mach einfach eine lokale Konstante, verschieben können wir sie immer noch.

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+
+	sort.Sort(RatesByPrice(data))
+
+	for i := 0; i < len(data); i++ {
```
das sollte gekapselt werden. Ist das nicht

    getCheapSlots && isCurrentSlotInList

schenlap replied:
ja, denke schon. Kann man machen

schenlap replied:
es hat da schon einige Abhängigkeiten. Ich habe alles rausgezogen was zusammen gehört https://github.com/schenlap/evcc/commit/8d805ccbd8245713a551d4979ccf39f5e2157697

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+				t.log.DEBUG.Printf("cheap timeslot, charging...\n")
+				t.cheapactive = true
+			} else {
+				if t.cheapactive && sum > duration+10*time.Minute {
```
wo kommt die Zeitkonstante her?

schenlap replied:
Hystherese. Wegen <10 Minuten abschalten macht keinen Sinn.

andig replied:
dan wär das gut als

    hysteresisDuration = 10*time.Minute

andig committed:
wip

schenlap committed:
kapsle

schenlap committed:
fix last timestamp check

andig commented on core/planner.go:
```diff
@@ -0,0 +1,167 @@
+	tariff      api.Tariff
+	mux         sync.Mutex
+	cheap       float64
+	cheapactive bool
```
Brauchts das als Instanzvariable?

schenlap replied:
cheapactive brauch ich. 

cheap könnte man lassen, dann kann es es verwenden wie bisher oder auch aus der Logik rausnehmen. Ich denke aber dass es Tarifmodelle gibt wo es Hoch / Niedertarif gibt. Da könnte es sinn machen die bisherige Logik verwenden zu wollen.

die anderen beiden sind für die Funktion notwendig.

andig replied:
> cheapactive brauch ich.

Wozu? Der Algorithmus ist simple und das läuft fix durch. Du verlässt Dich hier darauf, dass eine Variable gesetzt ist indem vorher eine andere Funktion aufgerufen sein muss. Das ist implizit und fehleranfällig. Ich würde die Berechnung lieber neu machen.

schenlap replied:
Ich brauche den Stand des letzten Durchlaufs. Ist beim Stundenwechsel relevant.

schenlap replied:
hab's wegbekommen indem ich im Timeslot immer möglichst bald beginne zu laden. Macht auch Sinn um reserve zu haben wenn es länger dauert als gedacht.

schenlap committed:
beginn test

schenlap committed:
tests: add IsCheap

schenlap committed:
simplify and add tests

schenlap committed:
remove cheapactive from instance variables

schenlap commented:
ich würde das Cheap komplett rauswerfen, spricht etwas dagegen? Ich sehe darin keinen Nutzen mehr.

schenlap committed:
rename GetFinishAt to GetTargetTime

andig commented:
> ich würde das Cheap komplett rauswerfen, spricht etwas dagegen? Ich sehe darin keinen Nutzen mehr.

Wir brauchen ein Signal an den Ladepunkt ob geladen werden soll. *was* günstig ist weiß der Planner aufgrund der Zielzeit. Es mag aber auch im PV Modus Situationen geben wo man gerne zusätzlich aus dem Netz bezieht, z.B. bei negativen Preisen. Den zweiten Fall deckt heute „isCheap“ ab. 

Bekommen wir Beides unter einen Hut?

schenlap commented:
>Bekommen wir Beides unter einen Hut?

aktuell geht beides. Dann müssen die von dir genannten Änderungen noch umgesetzt werden. (cheap aus konfig übernehmen, cheap in site verschieben). Auch da würde ich Unterstützung brauchen.

andig committed:
Move cheap rate to site

andig commented:
> > Bekommen wir Beides unter einen Hut?
> 
> aktuell geht beides. Dann müssen die von dir genannten Änderungen noch umgesetzt werden. (cheap aus konfig übernehmen, cheap in site verschieben). Auch da würde ich Unterstützung brauchen.

Hab ich eingebaut- Du bist wieder dran ;)

andig committed:
Add config

schenlap committed:
fix test

cheap rate is now in site

schenlap committed:
const

schenlap committed:
coding style

andig commented on core/loadpoint_api.go:
```diff
@@ -210,7 +210,6 @@ func (lp *LoadPoint) GetMaxPower() float64 {
 	return Voltage * lp.GetMaxCurrent() * float64(lp.GetPhases())
 }
 
-// setRemainingDuration sets the estimated remaining charging duration
```
revert

andig commented on core/loadpoint_api.go:
```diff
@@ -245,3 +244,13 @@ func (lp *LoadPoint) GetRemainingEnergy() float64 {
 	return lp.chargeRemainingEnergy
 }
+
+// GetTargetTime returns the charge finish time
```
// GetTargetTime returns the charge target time

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+
+	var cheapactive bool
+	if cheapSlotNow {
+		if curSlotNr == cntExpectedSlots { // delay most expensiv slot if not last slot
```
warum brauchen wir diese Logik? Ist es wirklich notwendig, einen Slot nochmal aufzuteilen?

schenlap replied:
ja, ist notwendig. Sonst passiert es dass zu Beginn des zweitgünstigsten Slot zu laden begonnen wird und im günstigsten dann vielleicht nur mehr 1 Minute über bleibt. Da kommt dann gleich ein Issue dass er nicht richtig geladen hat. Das will ich vermeiden. Besonders schlimm wird es bei Tarifen wo sich die Tarife nicht jede Stunde ändern sondern größere Blöcke gebildet werden. Haben wir noch nicht drinnen, gibt es aber auch und ist nur eine Frage der Zeit bis es jemand braucht.
Mit der eingestellten Chargeeffizienz wird das Problem teilweise verstärkt, kann ich aber nicht ändern.

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+}
+
+func (t *Planner) PlanActive(duration time.Duration, end time.Time) (bool, error) {
+	t.mux.Lock()
```
wofür hier die mutex?

schenlap replied:
nachdem wir jetzt Sort in jedem Durchlauf neu machen ist es vielleicht nicht mehr notwendig, schaue ich mir an

schenlap replied:
so wie er seit der Umstrukturierung in planner verwendet wird ist er nutzlos.
Da wir nur einmalig eine Kopie der Preise abholen müssen wir nur sicherstellen dass in dieser Zeit keine Änderung des Tariffs stattfindet. Wie funktioniert der Timer für das Update, ist der asynchron und können uns da Probleme machen?

andig replied:
Ok, das sollte kein Problem sein- die Tarife geben jedesmal ein neues Slice zurück.

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+			continue
+		}
+
+		if !(pi.Start.Before(end)) { // charge should ends before
```
if pi.Start.After(end) || pi.End.Before(t.clock.Now()) { ...

schenlap replied:
> @schenlap was fehlt denn noch um auf "Ready for review" zu kommen?

Ich denke wir sind schon mitten im Review :-), ich schalte um.

schenlap replied:
> if pi.Start.After(end) 

damit hatte ich Probleme da die Zeit während der Berechnung weiter fortschreitet.
After: ist ein >
!Before ist ein < bzw. >=, also nicht das selbe

Zusammenfassen in ein if geht natürlich

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+	sort.Sort(RatesByPrice(data))
+
+	for i := 0; i < len(data); i++ {
+		pi = data[i]
```
wofür steht pi?

schenlap replied:
PriceIntervall

andig replied:
periodPrice

oder 

    pData

Sonst hast Du ja auch überall p für `period`?

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+		}
+
+		// timeslot already started
+		pstart := pi.Start
```
ist das nicht doppelt zu zeile 82?

schenlap replied:
Ja

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+		}
+
+		// timeslot ends after charge finish time
+		pend := pi.End
```
besseres Naming:

    periodEnd

schenlap replied:
ich habs ganz rausgenommen, ist nicht mehr notwendig.

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+		}
+
+		cntExpectedSlots++
+		delta := pend.Sub(pstart)
```
delta variable weg

andig commented on core/planner.go:
```diff
@@ -0,0 +1,158 @@
+	}
+}
+
+func (t *Planner) isCheapSlotNow(duration time.Duration, end time.Time) (bool, error) {
```
Ich würde generell nochmal überlegen, ob nicht vllt. eine Funktion die Daten vorbereitet und start/end anpasst und die andere die tatsächliche Planung macht?

andig commented on core/soc/estimator.go:
```diff
@@ -174,6 +174,7 @@ func (s *Estimator) SoC(chargedEnergy float64) (float64, error) {
 			if !invalid && socDelta > 2 && energyDelta > 0 && s.prevSoC > 0 {
 				s.energyPerSocStep = energyDelta / socDelta
 				s.virtualCapacity = s.energyPerSocStep * 100
+				s.log.DEBUG.Printf("soc gradient updated: socDelta: %0.0fWh, energyDelta: %0.0fWh", socDelta, energyDelta)
```
was macht das hier?

schenlap replied:
ist wohl versehentlich reingerutscht bei Test von #2418 . Kommt wieder raus

schenlap replied:
f1d7ad4

andig commented on core/soc/timer.go:
```diff
@@ -52,6 +52,22 @@ func (lp *Timer) DemandValidated() bool {
+		power *= lp.current / lp.GetMaxCurrent()
+	}
+
+	remainingDuration := se.RemainingChargeDuration(power, lp.SoC)
```
return ...

andig commented on tariff/awattar.go:
```diff
@@ -7,22 +7,44 @@ import (
 
 var _ api.Tariff = (*Awattar)(nil)
 
+func fakeAwattar() awattar.Prices {
```
am Ende raus

schenlap replied:
klar

andig commented:
@schenlap was fehlt denn noch um auf "Ready for review" zu kommen?

schenlap committed:
simplify

schenlap committed:
remove fake function

schenlap committed:
rename

andig committed:
Simplify and clarfify naming

andig commented on core/planner.go:
```diff
@@ -0,0 +1,138 @@
+		}
+
+		// we found all necessary cheap slots to charge to targetSoC
+		if plannedDuration > requiredDuration {
```
>=

andig commented:
Ich hab nochmal aufgeräumt- die letzten Knackpunkte müsstest Du Dir bitte anschauen.

andig commented on core/planner.go:
```diff
@@ -0,0 +1,138 @@
+				t.log.DEBUG.Printf("continue charging in last slot\n")
+				cheapactive = true
+			} else { // expensiv and not last slot, delay
+				if plannedDuration > requiredDuration+hysteresisDuration {
```
Warum brauchen wir die Hysterese nochmal zusätzlich zur Effizienz? Ist das nicht ein bisschen viel Overkill? Ich wäre dafür die rauszuschmeissen und das _nur_ über Effizienz abzubilden.

schenlap replied:
ein zusätzlicher Puffer war eigentlich nicht der Grund für die Hysterese. Ich hatte Probleme dass mehrfach ein/ausgeschaltet wurde wenn sich der SoC des Autos gesprungen ist. Damals war es auch wirklich eine Hysterese, diesen Zweck erfüllt es aktuell so nicht mehr. Ich muss das in echt ohne dieser Hysterese nochmals testen um zu sehen ob es noch einen negativen Effekt hat wenn ich es weglasse.

andig replied:
@schenlap würdest Du da nochmal rein schauen? Vllt. einen Testcase für mit/ohne Hysterese bauen?

schenlap commented on core/planner.go:
```diff
@@ -0,0 +1,138 @@
+			slot.End = targetTime
+		}
+
+		cntExpectedSlots++
```
ja

schenlap commented on core/planner.go:
```diff
@@ -0,0 +1,138 @@
+	requiredDuration = time.Duration(float64(requiredDuration) / chargeEfficiency)
+
+	// Save same duration until next price info update
+	if targetTime.After(last) {
```
die bekannten Preise gehen nicht bis zum gewünschten Ladezeitpunkt. z.B am Freitag wird das Auto angesteckt und soll am Montag in der Früh abfahrbereit sein. Die Preise sind aber nur für Samstag verfügbar (bei Awattar ist das so). Jetzt will ich nicht die komplette Ladung am Samstag durchführen sondern ich teile es auf auf Samstag und behalte einen Rest für später. Sinn ist natürlich den PV Strom am Sonntag noch mitnehmen zu können. Das ist das am schlechtest getestete Feature weil mein Auto selten so lange steht.

schenlap committed:
fix test build

schenlap commented:
Jetzt produzieren alle Tests ein fail. Es wird nie eingeschaltet. Bei den aktuellen Strompreise wäre das vielleicht sogar sinnvoll :-) .
Auf den ersten Blick sind es nur Umbenennungen aber es hat sich grundlegend an der Funktion was geändert, das muss ich zuerst finden bevor ich weitermachen kann.

schenlap commented:
Das wäre eine mögliche Lösung. Es wird zuvor auf t.clockNow gesetzt, da die Zeit noch nicht weitergeschritten ist, wird Before nicht erfüllt. Ev. muss man es auch anders lösen.
```
@@ -101,7 +104,7 @@ func (t *Planner) PlanActive(requiredDuration time.Duration, targetTime time.Tim
                        slot.Price, plannedDuration)
 
                // current timeslot is a cheap one
-               if slot.Start.Before(t.clock.Now()) && slot.End.After(t.clock.Now()) {
+               if !slot.Start.After(t.clock.Now()) && slot.End.After(t.clock.Now()) {
                        cheapSlotNow = true
                        curSlotNr = cntExpectedSlots
                        t.log.TRACE.Printf(" (now, slot number %v)", curSlotNr)

```

schenlap committed:
remove logging to simplify deleayed start block

schenlap commented on core/planner.go:
```diff
@@ -0,0 +1,138 @@
+			} else { // expensiv and not last slot, delay
+				if plannedDuration > requiredDuration+hysteresisDuration {
+					t.log.DEBUG.Printf("cheap timeslot, delayed for %s\n", (plannedDuration - requiredDuration).Round(time.Minute))
+					cheapactive = false
```
Ja das läßt sich stark vereinfachen wenn man auf das Logging verzichtet. Habe geändert mit 7edeab7

schenlap commented on core/site_test.go:
```diff
@@ -25,4 +25,5 @@ func TestSitePower(t *testing.T) {
 	}
 }
 
+// TODO add test case for cheapRate
```
es wird aktuell nicht mehr getestet ob es auch Lädt wenn der Tarif unter dem eingestellten Tarif in der Config liegt. Zuvor war die Abfrage in planner drinnen und wurde auch getestet. In site.got gibt es leider so viele Abhängigkeiten dass das nicht einfach zum Testen ist (zumindest nich für mit mit wenig go und mock Erfahrung). Sinnvoll wäre das Test aber schon.

andig committed:
wip

andig commented:
Es bleibt noch:

```
    planner_test.go:111: fixed tariff case 1: expected true, got false
    planner_test.go:111: fixed tariff case 1: expected true, got false
FAIL
```

schenlap commented:
Mit 
`if !slot.Start.After(t.clock.Now())`
bekomme ich keinen Fail. Mit dieser Änderung bekommst du
```
    planner_test.go:111: fixed tariff case 1: expected true, got false
    planner_test.go:111: fixed tariff case 1: expected true, got false
```
?

schenlap commented:
Ich glaube ich habe noch eine große Unschönheit gefunden:
Ist kein Tarif konfiguriert, dann wird mit Zielzeit über timer.go geladen.
Konfiguriert man einen fixed-Tarif, z.B wegen der Ersparnisanzeige, dann wird mit planner.go geladen.
Nur weil mein einen Preis einstellen will darf sich die Ladelogik nicht ändern. Auch wenn beide das selbe Verhalten erzeugen ist es nicht schön wenn unterschiedliche Code-Teile ausgeführt werden.
Wie ich es in awattar implementiert habe, gab es das Problem nicht. War aber auch nicht sauber. Mit der Umstellung auf planner ist da jetzt der fixed Tarif aber mit rein gerutscht.

andig committed:
Simplify cheapactive

andig committed:
simpler

andig commented:
> Wie ich es in awattar implementiert habe, gab es das Problem nicht. War aber auch nicht sauber. Mit der Umstellung auf planner ist da jetzt der fixed Tarif aber mit rein gerutscht.

Mhhm. Für's erste könnten wir den Typ prüfen und den ausschließen. Aber wovon hängt das ab? Ob der Tarif überhaupt nur 1 Zone hat? Was bedeutet das mit Bezug auf https://github.com/evcc-io/evcc/issues/2524?

andig committed:
Fix test

schenlap commented:
Typ prüfen ist auch nicht schön. planner sollte timer.go eigentlich komplett ersetzen. Wobei das jetzt nicht so ist (nämlich dann wenn der SoC zum Zielzeitpunkt noch nicht erreicht wird, übernimmt derzeit noch der timer die restliche Ladung)

andig commented:
> Mit `if !slot.Start.After(t.clock.Now())` bekomme ich keinen Fail. Mit dieser Änderung bekommst du
> 
> ```
>     planner_test.go:111: fixed tariff case 1: expected true, got false
>     planner_test.go:111: fixed tariff case 1: expected true, got false
> ```
> 
> ?
> 
> Das wäre dann vielleicht wirklich ein Fehler im Test selbst. Muss ich mir ansehen

Der "Fehler" ist in den Tests und im Verständnis von before/after und der Frage ob die genau matchende Millisekunde dazu zählt. Da das in der Realität irrelevant ist sollten m.E. die Tests defensiv sein und die Timestamps um 1ns verschieben, im Code selbst sollten wir das nicht tun.

andig commented:
> Typ prüfen ist auch nicht schön. planner sollte timer.go eigentlich komplett ersetzen. Wobei das jetzt nicht so ist (nämlich dann wenn der SoC zum Zielzeitpunkt noch nicht erreicht wird, übernimmt derzeit noch der timer die restliche Ladung)

Das eine ist doch eine Zeitsteuerung, das andere eine Preissteuerung. Preissteuerung macht nur Sinn wenn es unterschiedliche Preise gibt (>1 Zone, >1 Preis). 

Alternativ: wir schmeissen die Zeitsteuerung weg und ersetzen sie- sollte es keinen Grid Tarif geben- durch einen "fake" Grid Tarif mit Preis 0.

andig committed:
ignore fixed tariff

andig commented:
Ich frag mich auch grad, warum eigentlich die `site` bestimmen sollte, ob der Ladepunkt laden muss wg. seiner Planung. Auch nicht schön...

schenlap committed:
equal or greater

schenlap committed:
test

andig commented:
Bingo! Im Ladepunkt können wir die Prio der beiden Planner umdrehen -> Problem gelöst.  Also erst Restzeit, dann nach Tarif.

andig commented on core/planner_test.go:
```diff
@@ -87,8 +87,8 @@ func TestIsCheapSlotNow(t *testing.T) {
 		{"fixed tariff", []float64{2}, 5 * time.Hour, []se{
-			{1, 0, 2 * time.Hour, true},
-			{1, 0, 10 * time.Minute, true},
+			{1, time.Minute, 2 * time.Hour, true},
```
1 wie 1ns hätte auch gereicht

schenlap replied:
soll ich nochmal änddern?

schenlap replied:
war sowieso noch ein Fehler drinnen :-)

schenlap committed:
test

schenlap committed:
remove debug output

andig committed:
Move planner to loadpoint step 1

andig committed:
Handle nil

andig commented:
Update: sorry für die Verzögerung. Der Komplettumbau erfordert einige Anpassungen am bestehenden Zielladen die etwas komplexer sind. Bin noch dran...

andig committed:
Fix test

andig commented:
Alternativ könnten wir auch die aktuelle Variante schonmal mergen und den Umbau in Schritt 2 machen?

schenlap committed:
Merge branch 'evcc-io:master' into dynamic_pricing

andig commented:
Depends on https://github.com/evcc-io/evcc/pull/2780 which will provide better abstraction for the time planner.

stefaanbolle commented:
My Electricity provider publishes the daily tariff using an API in JSON format. Is HTTP as source type supported for tariff parsing?If not, it might be an option to allow type: custom and source: http for using the feedin and grid tariffs

andig commented:
Bitte neues Thema- wer ist Dein Anbieter?

stefaanbolle commented:
> 

Done in discussion. Topic #3333
Provider is Eneco

schenlap commented:
> Update: sorry für die Verzögerung. Der Komplettumbau erfordert einige Anpassungen am bestehenden Zielladen die etwas komplexer sind. Bin noch dran...

@andig 
Da der Winter wieder naht und damit die PV Erzeugung weniger frage ich mal vorsichtig nach. Besteht hier von deiner Seite noch Interesse dies reinzubringen? Ist der Unterbau sowei umgebaut das man da weitermachen könnte?

andig commented:
Interesse besteht absolut, nur Zeit ist knapp. Ich will das auf jeden Fall umsetzen wenn ein bisschen Luft ist!!!

afuxi commented:
Nutze es gern, wenn es kommt.

andig commented:
@schenlap ich würde mir das gerne genauer anschauen.

andig commented:
@schenlap weiter gehts in https://github.com/evcc-io/evcc/pull/5445. Ich habe diesen PR rebased und das `Tariff` Api etwas vereinfacht. Next step: soc timer fixen.
